File: app/src/androidTest/java/com/adsamcik/signalcollector/utility/TranslateableStringTest.java
package com.adsamcik.signalcollector.utility;


import android.content.Context;
import android.support.annotation.NonNull;
import android.support.test.InstrumentationRegistry;

import com.adsamcik.signalcollector.R;
import com.google.gson.Gson;

import junit.framework.Assert;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

@RunWith(JUnit4.class)
public class TranslateableStringTest {
	private Context appContext = InstrumentationRegistry.getTargetContext();
	private Gson gson = new Gson();
	private final String identifier = "activity_idle";
	private final int identifierInt = R.string.activity_idle;
	private final String defaultString = "Default string";

	@Test
	public void loadFromJson() {
		final String json = "{\"defaultString\":\"" + defaultString + "\",\"identifier\":" + identifier + "}";
		Assert.assertEquals(appContext.getString(identifierInt), gson.fromJson(json, TranslateableString.class).getString(appContext));
	}

	@Test
	public void basicValueTest() {
		final TranslateableString target = new TranslateableString(identifier, defaultString, (identifier) -> identifierInt);

		Assert.assertEquals(defaultString, target.getDefaultString());
		Assert.assertEquals(identifier, target.getIdentifier());
		Assert.assertEquals(appContext.getString(identifierInt), target.getString(appContext));
	}

	@Test
	public void exceptionTest() {
		final TranslateableString target = new TranslateableString();

		try {
			target.getString(appContext);
			throw new IllegalStateException("Get string did not throw exception");
		} catch (RuntimeException e) {
			//everything works as expected
		}

		target.setIdentifier("");

		try {
			target.getString(appContext);
			throw new IllegalStateException("Get string did not throw exception");
		} catch (RuntimeException e) {
			//everything works as expected
		}
	}

	@Test
	public void dynamicResourceTest() {
		final TranslateableString target = new TranslateableString();
		target.setIdentifier(identifier);
		Assert.assertEquals(appContext.getString(identifierInt), target.getString(appContext));
	}
}

File: app/src/androidTest/java/com/adsamcik/signalcollector/utility/TranslateableStringTest.kt
package com.adsamcik.signalcollector.utility


import android.support.test.InstrumentationRegistry
import com.google.gson.Gson
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4

@RunWith(JUnit4::class)
class TranslateableStringTest {
    private val appContext = InstrumentationRegistry.getTargetContext()!!
    private val gson = Gson()
    private val identifier = "activity_idle"
    private val identifierInt = R.string.activity_idle
    private val defaultString = "Default string"

    @Test
    fun loadFromJson() {
        val json = "{\"defaultString\":\"$defaultString\",\"identifier\":$identifier}"
        Assert.assertEquals(appContext.getString(identifierInt), gson.fromJson(json, TranslateableString::class.java).getString(appContext))
    }

    @Test
    fun basicValueTest() {
        val target = TranslateableString(identifier, defaultString,
                object : TranslateableString.IIdentifierResolver {
                    override fun resolve(identifier: String): Int = identifierInt
                })

        Assert.assertEquals(defaultString, target.defaultString)
        Assert.assertEquals(identifier, target.identifier)
        Assert.assertEquals(appContext.getString(identifierInt), target.getString(appContext))
    }

    @Test
    fun exceptionTest() {
        val target = TranslateableString()

        try {
            target.getString(appContext)
            throw IllegalStateException("Get string did not throw exception")
        } catch (e: RuntimeException) {
            //everything works as expected
        }

        target.identifier = ""

        try {
            target.getString(appContext)
            throw IllegalStateException("Get string did not throw exception")
        } catch (e: RuntimeException) {
            //everything works as expected
        }

    }

    @Test
    fun dynamicResourceTest() {
        val target = TranslateableString()
        target.identifier = identifier
        Assert.assertEquals(appContext.getString(identifierInt), target.getString(appContext))
    }
}
